### DISCUSSION TITLE
Potential Foxglove support for FRC (WPILib)

### DISCUSSION THREAD

**Discussion by jkuszmaul** [foxglove/community#46] [Q&A]
Note: This is separate from any AOS discussions related to the frc971 repo, which I've instigated in the past. This is about how we might use Foxglove to provide data visualization for a wider swath of teams.

@peterjohnson has been looking at how to improve data visualization for the typical FRC team. One of the things that came up is Foxglove, as it already provides some potentially useful features. However, in order to make the integration/workflow as smooth as possible for FRC teams and to provide some of the features that @peterjohnson is looking for, Foxglove would need to support some new things, of potentially varying complexity.

There are some notes of desirable new features for visualization at https://github.com/wpilibsuite/allwpilib/projects/14

The biggest potential change that we were discussing is support for multiple data sources. Currently FRC teams may have data from:
* On the robot
* The driver-station logs
* A match video taken from a phone/camera in the stands

And then wish to see those all in one place, potentially with the ability to dynamically mess with the time offsets between the data sources to be able to get them aligned. This sounds like a pretty substantial feature request to me, but I'm not all that familiar with how things work.

Other changes that seem less drastic:
* Figuring out how to support the FRC data sources (whether by supporting them natively in Foxglove, or ensuring that the FRC data sources are in the right format, or by making some easy-to-run converter).
* Support for live plotting of whatever format we are using (or adopt the websocket protocol).
* Support for sending network tables values.

Note that NetworkTables, which is what WPILib provides to do communication between network nodes. It is slated to change substantially this off-season, so there may be opportunity to ensure that the protocol is appropriate for interfacing with studio.

I personally am not as involved with the WPILib development, so will let the rest of you figure out if I've communicated anything poorly.

**Comment by jtbandes** [comment]
Thanks for kicking off the discussion! Here are some high level thoughts:

There are a few paths to getting data into Foxglove Studio.
- First-class support for multiple data sources is tricky (but is also something we've seen some prior requests for: https://github.com/foxglove/studio/issues/2064, https://github.com/foxglove/studio/issues/3174).
- We've leaned in the direction of supporting some common formats that users can translate their data into, rather than inviting users to write and maintain their own data source plugins, which could be subject to more internal API changes in the app.
  - ([Custom panel extensions](https://foxglove.dev/docs/studio/extensions/getting-started) can fetch data from anywhere they like, but the extension API doesn't provide a way to plumb that data back into the rest of the app.)
  - For recorded data, we support the [MCAP](https://mcap.dev) file format which can contain schemas & data in any encoding (and we can extend Foxglove Studio with support for new encodings if desired). Perhaps JSON would be sufficient for FRC/NetworkTables use cases?
  - For live connections, we support a [WebSocket-based server protocol](https://foxglove.dev/docs/studio/connection/foxglove-websocket) that is similar to MCAP in spirit, but has dynamic operations for the server to dynamically add/remove data streams. (Publishing data back to the server is not yet implemented, but is [a common request](https://github.com/foxglove/ws-protocol/issues/38) that we would definitely like to support.)
- One option would be to use a single "bridge" process that speaks WebSocket to Foxglove Studio, and on the other end interfaces with various data sources as necessary.
- On the other hand, a natively-supported NetworkTables connection could make sense, kind of like our [native ROS connections](https://foxglove.dev/docs/studio/connection/native) and [Velodyne LIDAR connection](https://foxglove.dev/docs/studio/connection/velodyne). This could live in the studio repo and wouldn't require users to run a custom "bridge" process alongside their other software components.
  - Any connection that requires custom TCP/UDP support (not just HTTP/WebSockets) would be restricted to our desktop app, and wouldn't work in the browser version at https://studio.foxglove.dev.

Addressing some of the other tickets on the Sapphire project you linked:
- Our app is pretty heavily indexed on a "topics"-based workflow (where each topic is a named stream that contains messages conforming to the same schema). If NetworkTables users are used to seeing a "tree"-like view, it might take some thinking for how to expose these nicely in the app. [Custom panels](https://foxglove.dev/docs/studio/extensions/getting-started) are an option for sure, but it seems best to try and take advantage of the app's built-in features as much as possible.
- Timeline events are something we are actively working on building into our [data platform product](https://foxglove.dev/docs/data-platform/api/events). This means we will also eventually add support to Studio for visualizing and possibly editing events.
- Our [plot support](https://foxglove.dev/docs/studio/panels/plot) is somewhat advanced but we do know that we're reaching some performance limitations with the current architecture and renderer, so we are interested in exploring alternatives (although this isn't currently a high priority for us).

There are a lot of things in the linked board that we don't know anything about (HAL websockets?), so please feel free to expound here if we're missing anything!

**Reply by jkuszmaul** [reply]
FYI the NetworkTables changes I alluded to include making it more of a pubsub system, so it should have the concepts necessary to fit cleanly with Foxglove.

I'll let Peter worry about whether to write bridges, or push for native support of a given protocol at either end. For AOS stuff we haven't minded just running a separate bridge, but I don't have quite as good a feel for the constraints for the regular WPILib. I know there was some concern that the existing WPILib team has minimal web/typescript experience and so might but be and to contribute as much as they might like to .

I will say that for improved plotting features, that's something I'd like to see independent of this effort, since a substantial proportion of the visualization I do is with 2d plots of varying complexity. On the other hand, the current plots are decent and fulfill the bulk of my needs. Once I've had some chance to actually use Foxglove more (rather than just getting it integrated), I'll try to make some feature requests and/or use that as a jumping board to try to contribute some myself.

**Comment by PeterJohnson** [comment]
I think the critical must-have feature for the FRC use case would be the ability to load multiple data sources and tweak time alignment between them within Foxglove.  While it would be possible to write a separate tool to pre-process multiple data sources into one, the workflow would be terrible to time-align them with this approach--it would be an iterative process to merge, load Foxglove, look at the time deltas, re-merge / re-load, etc.  Particularly when there's things like video sources where it may not be immediately obvious where to sync them.

The rest of the feature requests are things we could likely implement ourselves with some moderate assistance (e.g. bridges to FRC specific protocols, loading our data formats, improvements to plots, etc).

**Reply by jtbandes** [reply]
Could you describe more about the different data sources that would need to be loaded at once?

**Reply by PeterJohnson** [reply]
First, a brief description of our system configuration:
- Robot controller: an ARM-based microprocessor running Linux
- Driver station: Windows PC with joysticks attached.  Runs a real-time Windows application (also called the Driver Station, or DS) written in LabVIEW that reads joysticks and sends control signals (enable, joystick info) to the robot controller.  Teams can also run dashboards - separate Windows applications to visualize real-time data coming from the Robot controller (and can send commands/data changes as well)
- Coprocessors: separate microprocessors on the robot, used typically for vision processing (OpenCV).  Communicate with both the robot controller and dashboards
- The robot has an Ethernet switch to interconnect the robot controller and coprocessors and a WiFi bridge.  The driver station connects via WiFi to the robot.
- Teams of high school students write custom robot programs (languages include Java, C++, Python, and LabView) that run on the robot controller.  Some teams write custom coprocessor programs or dashboard applications, but semi-standard ones are provided to teams by us or 3rd party vendors.

Potential sources of data:
1. Data log (custom binary or CSV) captured on robot controller; possibly multiple logs in the case when the robot code or controller restarts in the middle of a match.
2. Data log (and potentially separate video files) captured on coprocessor(s)
3. Data log captured on driver station PC by Driver Station application (LabVIEW custom format)
4. Video captured on driver station PC (of field, and/or streamed from robot)
5. Video file captured on cellphone in stands (or capture of field online video stream broadcast via Twitch)

Most operating periods on the robot are in the range of 10-20 ms.  Currently none of these data sources are time-synchronized and don't share common timestamps or time bases.  We're likely on a path where 1 and 2 are synchronized (potentially except in the case of robot reboot).  We have a pretty good common trigger for synchronizing 3 with the others within ~20 ms (enable signal).  Things like video are harder to synchronize--you pretty much have to look at the image and tweak to match other events in some other log (e.g. seeing the robot arm start to move and match that to a sensor change on the robot).  The "data logs" here have similarity to ROS bags--they consist of multiple named timestamped data value streams of various formats (integers, doubles, strings, arrays of the above, raw binary, structured formats).

**Reply by jkuszmaul** [reply]
fwiw, since to my understanding there is already an NTP server running on the driver station that the roborio (robot controller) listens to, we should be able to do much better than 20 ms for (1) <-> (3) time synchronization. But I've never actually benchmarked that particular synchronization.

**Reply by jkuszmaul** [reply]
But it does seem like it may be feasible to make it so that the only thing which might need user-specified time synchronization would be basically just video files. And since videos will typically only need to be shown in a single panel on their own (*maybe* with some annotations, but for spectator video not sure even that is needed), and because of that we could make it so that we just define a custom extension specifically for adding on video streams with a time offset slider, rather than needing to treat them as first-class data sources, which limits the scope of the architectural changes required to support dynamic time offsets.

**Reply by PeterJohnson** [reply]
It's not as sophisticated as a NTP server--the DS simply sends the datetime of the laptop to the robot controller via UDP periodically and there's a service there (netcomm) that sets the local time.  Timestamp wise we use FPGA time (1 us) instead of local time, although it is possible to sync the two fairly closely (the current log file format dumps a FPGA-timestamped local time every 5 seconds or so for this exact reason).
